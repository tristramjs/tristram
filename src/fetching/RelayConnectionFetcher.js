/* @flow */
import type { RawSiteMapData } from '../types/sitemap';

import type { Fetcher, FetcherProps } from './index';

require('es6-promise').polyfill();
require('isomorphic-fetch');

type Props<T> = FetcherProps<T> & {
	chunkSize?: number,
	getConnection: GetConnection<T>,
	query: string,
	logErrors?: boolean,
	maxRetries?: number,
};

type GetConnection<T> = <T>(x: Object) => GraphQlConnection<T>;
type GraphQlConnection<T> = { edges: { node: T, cursor: string }[], pageInfo: { hasNextPage: boolean } };

export default class RelayConnection<T> implements Fetcher {
	url: string;
	query: string;
	chunkSize: number;
	transformResult: T => RawSiteMapData;
	logErrors: boolean;
	maxRetries: number;
	getConnection: GetConnection<T>;
	variables: {
		first: number,
		after?: string,
	} = {
		first: 10,
	};

	constructor({
		url, getConnection, transformResult, query, chunkSize, logErrors, maxRetries,
	}: Props<T>) {
		this.url = url;
		this.getConnection = getConnection;
		this.transformResult = transformResult;
		this.query = query;
		this.maxRetries = maxRetries || 3;
		this.chunkSize = chunkSize || 100;
		this.logErrors = logErrors || false;
	}

	async* getData(): AsyncGenerator<RawSiteMapData[], void, void> {
		let hasNextPage;

		do {
			const data: GraphQlConnection<T> = await this.fetchConnection();

			yield data.edges.map(edge => this.transformResult(edge.node));
			({ hasNextPage } = data.pageInfo);

			this.variables = {
				first: this.chunkSize,
				after: data.edges[data.edges.length - 1].cursor,
			};
		} while (hasNextPage);
	}

	async fetchConnection(): Promise<GraphQlConnection<T>> {
		const {
			url, query, variables, getConnection, logErrors,
		} = this;
		let retries = 0;
		const options = {
			method: 'POST',
			headers: {
				'content-type': 'application/json',
			},
			body: JSON.stringify({ query, variables }),
		};

		while (retries < this.maxRetries) {
			const res = await fetch(url, options);
			const data = await res.json();

			if (data.errors && Array.isArray(data.errors)) {
				if (logErrors) {
					// eslint-disable-next-line no-console
					console.log(
						`RelayConnectionFetcher: Received errors in connection:\n"${data.errors
							.map(err => (err.message ? err.message : '<no error message provided>'))
							.join('\n')}"`
					);
				}

				retries = retries + 1;
			} else {
				return getConnection(data);
			}
		}

		throw new Error(`Error fetching connection, tried ${retries} times.`);
	}
}
