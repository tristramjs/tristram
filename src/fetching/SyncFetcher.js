/* @flow */
import type { RawSiteMapData } from '../types/sitemap';

import type { ChunkFetcher } from './index';

export default class SyncFetcher implements ChunkFetcher {
	data: RawSiteMapData | RawSiteMapData[];

	constructor({ data }: { data: RawSiteMapData | RawSiteMapData[] }) {
		this.data = data;
	}

	async* getData(): AsyncIterator<RawSiteMapData[]> {
		if (Array.isArray(this.data)) {
			yield this.data;
		} else {
			yield [ this.data ];
		}
	}
}
