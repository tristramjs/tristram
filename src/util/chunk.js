/* @flow */

function chunk<T>(arr: T[], chunkSize: number): T[][] {
	const chunks = [];
	let index = 0;
	const { length } = arr;

	while (index < length) {
		chunks.push(arr.slice(index, (index += chunkSize)));
	}

	return chunks;
}

export default chunk;
