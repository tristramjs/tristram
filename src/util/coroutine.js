/* @flow */

/**
 * Returns a function that, when called,
 * returns a generator object that is immediately
 * ready for input via `next()`
 */
export default function coroutine(generatorFunction: () => AsyncGenerator<*, *, *>) {
	return function cb(...args: any) {
		const generatorObject = generatorFunction(...args);
		generatorObject.next();
		return generatorObject;
	};
}
