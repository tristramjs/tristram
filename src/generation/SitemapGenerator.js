/* @flow */
import type { Fetcher } from '../fetching';
import type { Writer } from '../persistence';
import type { Formatter } from '../formatting';
import type { RawSiteMapData } from '../types/sitemap';
import chunk from '../util/chunk';

// options management? We have several Modules who needs to know about options
export type Options = {
	hostname: string,
	publicPath?: string,
	cacheTime?: number,
	maxItemsPerSitemap?: number,
	maxItemsPerIndexSitemap?: number,
};

type Props = {
	fetchers: Fetcher[],
	// formatter types..
	formatter: Formatter,
	options: Options,
	writer: Writer,
};

export type OptionsWithDefaults = {
	hostname: string,
	cacheTime?: number,
	maxItemsPerSitemap: number,
	maxItemsPerIndexSitemap: number,
};

type Sitemap = string;

export default class SitemapGenerator {
	fetchers: Fetcher[];
	formatter: Formatter;
	options: OptionsWithDefaults;
	writer: Writer;
	sitemaps: Sitemap[];
	publicPath: string;

	buffer = [];

	constructor({
		fetchers, formatter, options, writer,
	}: Props) {
		this.fetchers = fetchers;
		this.formatter = formatter;
		this.writer = writer;
		this.sitemaps = [];
		this.publicPath = options.publicPath || '';

		this.options = {
			maxItemsPerSitemap: 50000,
			maxItemsPerIndexSitemap: 50000,
			...options,
		};
	}

	async run() {
		for (const fetcher of this.fetchers) {
			for await (const data of fetcher.getData()) {
				await this.saveChunk(data);
			}
		}

		if (this.buffer.length > 0) {
			await this.writeSitemap(this.buffer);
			this.buffer = [];
		}

		await this.writeIndexSitemap();
	}

	async saveChunk(data: RawSiteMapData[]) {
		const { options: { maxItemsPerSitemap } } = this;
		if (this.buffer.length + data.length < maxItemsPerSitemap) {
			this.buffer = this.buffer.concat(data);
		} else if (this.buffer.length + data.length === maxItemsPerSitemap) {
			await this.writeSitemap(this.buffer.concat(data));
			this.buffer = [];
		} else if (this.buffer.length + data.length > maxItemsPerSitemap) {
			// 1. get a chunk of the new data such that (currentItemCount + dataChunk.length == maxItemsPerSitemap)
			// so it can be flushed out to the writer
			const { length: bufferLength } = this.buffer;
			const firstChunk = data.slice(0, maxItemsPerSitemap - bufferLength);
			await this.writeSitemap(this.buffer.concat(firstChunk));
			this.buffer = [];

			// 2. for all chunks where (chunk.length == maxItemsPerSitemap), flush out to Writer
			const [ lastChunk, ...otherChunks ] = chunk(
				data.slice(maxItemsPerSitemap - bufferLength),
				maxItemsPerSitemap
			).reverse();

			for (const part of otherChunks) {
				await this.writeSitemap(part);
			}

			// 3. if the final chunk is (finalChunk.length < maxItemsPerSitemap),
			// create a new sitemap and flush the (incomplete) data chunk ELSE just save the complete sitemap
			if (lastChunk.length < maxItemsPerSitemap) {
				this.buffer = this.buffer.concat(lastChunk);
			} else if (lastChunk.length === maxItemsPerSitemap) {
				await this.writeSitemap(lastChunk);
			}
		}
	}

	async writeSitemap(data: RawSiteMapData[]) {
		const { xmlDeclaration, openingTag, closingTag } = this.formatter;
		const body = this.formatter.format(data);
		const sitemap = [ xmlDeclaration, openingTag, body, closingTag ].join('');

		const map = await this.writer.write(`sitemap-${this.sitemaps.length}.xml`, sitemap);
		this.sitemaps.push(map);
	}

	async writeIndexSitemap() {
		const numberIndexSitemaps = Math.ceil(this.sitemaps.length / this.options.maxItemsPerIndexSitemap);
		const { hostname } = this.options;
		const numberSitemaps = this.sitemaps.length;

		for (let index = 0; index < numberIndexSitemaps; index = index + 1) {
			const indexSitemap = this.formatter.formatIndex({ path: this.publicPath, hostname, numberSitemaps });

			await this.writer.write('index-sitemap.xml', indexSitemap);
		}
	}
}
