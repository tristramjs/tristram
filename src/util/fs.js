/* @flow */
import fs from 'fs';
import util from 'util';

const appendFile = util.promisify(fs.appendFile);
const exists = util.promisify(fs.exists);
const mkdir = util.promisify(fs.mkdir);
const readdir = util.promisify(fs.readdir);
const readfile = util.promisify(fs.readFile);
const unlink = util.promisify(fs.unlink);
const rmdir = util.promisify(fs.rmdir);
const writeFile = util.promisify(fs.writeFile);

export { appendFile, exists, mkdir, readdir, readfile, unlink, rmdir, writeFile };
