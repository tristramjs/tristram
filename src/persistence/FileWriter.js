/* @flow */
import { appendFile, writeFile } from '../util/fs';

import type { Writer, Path } from './index';

type Props = {
	path: Path,
	fileName: string,
};

export default class FileWriter implements Writer {
	path: Path;
	fileName: string;

	constructor({ path, fileName }: Props) {
		this.path = path;
		this.fileName = fileName;
	}

	async createSitemap(xmlDeclaration: string, openingUrlSet: string, partNumber: number): Promise<Path> {
		const path = this.getSitemapPath(partNumber);
		await writeFile(path, xmlDeclaration + openingUrlSet);

		return path;
	}

	async writeChunk(data: string, partNumber: number) {
		await appendFile(this.getSitemapPath(partNumber), data);
	}

	async commitSitemap(closingUrlSet: string, partNumber: number) {
		await appendFile(this.getSitemapPath(partNumber), closingUrlSet);
	}

	async createIndexSitemap(data: string) {
		await writeFile(`${this.path}indexSitemap.xml`, data);
	}

	getSitemapPath(partNumber: number) {
		return `${this.path}${this.fileName}-${partNumber}.xml`;
	}
}
