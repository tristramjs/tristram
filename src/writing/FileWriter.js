/* @flow */
import { appendFile, writeFile } from '../util/fs';

import type { Writer, Path } from './index';

type Props = {
	path: Path,
};

export default class FileWriter implements Writer {
	path: Path;
	fileName: string;

	constructor({ path }: Props) {
		this.path = path;
	}

	async write(key: string, data: string) {
		const path = this.path + key;
		await writeFile(path, data);
		return path;
	}
}
