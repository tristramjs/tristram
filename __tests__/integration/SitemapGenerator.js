/* @flow */

import { setup, cleanup } from '../__testHelpers__/fs';
import { readdir, readfile } from '../../src/util/fs';
import SitemapGenerator from '../../src/generation/SitemapGenerator';
import SyncFetcher from '../../src/fetching/SyncFetcher';
import RelayConnectionFetcher from '../../src/fetching/RelayConnectionFetcher';
import FileWriter from '../../src/writing/FileWriter';
import PlainFormatter from '../../src/formatting/Plain';

const path = `${process.cwd()}/MainIntegrationTest/`;

describe('SitemapGenerator module', () => {
	beforeAll(setup(path));

	it('should run without errors', async () => {
		const options = { hostname: 'foo.bar', maxItemsPerSitemap: 2 };

		const syncFetcher = new SyncFetcher({
			data: [ { loc: 'quz' }, { loc: 'qaz' }, { loc: 'qak' } ],
		});

		const chunkFetcher = new RelayConnectionFetcher({
			url: 'http://test.com/graphql',
			query: /* GraphQL */ `query Test($first: Int!, $after: String) {
				viewer {
					allTestItems(first: $first, after: $after) {
						pageInfo {
							hasNext
						}
						edges {
							cursor
							node {
								id
							}
						}
					}
				}
			}`,
			getConnection: conn => conn.data.viewer.allTestItems,
			transformResult: (data) => {
				const returnData = { ...data };
				returnData.loc = data.id;
				delete returnData.id;
				return returnData;
			},
		});

		const main = new SitemapGenerator({
			fetchers: [ syncFetcher, chunkFetcher ],
			formatter: new PlainFormatter(),
			writer: new FileWriter({ path }),
			options,
		});

		await main.run();

		const files = await readdir(path);
		for (const file of files) {
			expect(await readfile(`${path}/${file}`, 'utf8')).toMatchSnapshot();
		}

		expect(files.length).toBe(5);
	});

	afterAll(cleanup(path));
});
