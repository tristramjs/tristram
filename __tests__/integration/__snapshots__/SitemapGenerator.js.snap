// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SitemapGenerator module should run without errors 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><sitemapindex xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><sitemap><loc>https://foo.bar/sitemap-2.xml</loc></sitemap><sitemap><loc>https://foo.bar/sitemap-1.xml</loc></sitemap><sitemap><loc>https://foo.bar/sitemap-0.xml</loc></sitemap></sitemapindex>"`;

exports[`SitemapGenerator module should run without errors 2`] = `"<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>quz</loc></url><url><loc>qaz</loc></url></urlset>"`;

exports[`SitemapGenerator module should run without errors 3`] = `"<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>qak</loc></url><url><lastmod>2009-11-05T11:20:30.000Z</lastmod><priority>0.5</priority><changefreq>never</changefreq><loc>one</loc></url></urlset>"`;

exports[`SitemapGenerator module should run without errors 4`] = `"<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>two</loc><image:image><image:loc>http://example.com/image.jpg</image:loc><image:caption>an image</image:caption><image:geo_location>Earth, Milkyway</image:geo_location><image:title>Example</image:title><image:license>MIT</image:license></image:image><image:image><image:loc>http://example.com/photo.jpg</image:loc></image:image></url><url><loc>three</loc><video:video><video:thumbnail_loc>http://www.example.com/thumbs/123.jpg</video:thumbnail_loc><video:title>Grilling steaks for summer</video:title><video:description>Alkis shows you how to get perfectly done steaks every time</video:description><video:content_loc>http://www.example.com/video123.mp4</video:content_loc><video:player_loc>http://www.example.com/videoplayer.mp4?video=123</video:player_loc><video:duration>600</video:duration><video:expiration_date>2009-11-05T11:20:30.000Z</video:expiration_date><video:rating>4.2</video:rating><video:view_count>12345</video:view_count><video:publication_date>2007-11-05T11:20:30.000Z</video:publication_date><video:family_friendly>yes</video:family_friendly><video:tag>This</video:tag><video:tag>is</video:tag><video:tag>a</video:tag><video:tag>tag</video:tag><video:restriction relationship=\\"deny\\">GB US</video:restriction><video:gallery_loc>http://bpaul.us</video:gallery_loc><video:price currency=\\"EUR\\">1</video:price><video:price currency=\\"USD\\">2</video:price><video:uploader>Mr Mister</video:uploader><video:requires_subscription>yes</video:requires_subscription><video:live>no</video:live><video:platform relationship=\\"deny\\">tv</video:platform></video:video></url></urlset>"`;

exports[`SitemapGenerator module should run without errors 5`] = `"<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>four</loc></url></urlset>"`;
